### PR TITLE
Train only discs landmarks

### PR DESCRIPTION
**Pull request by yw7**
To fix errors in the vertebrae segmentation we train model using only disc landmarks:

image

We also update the model description with this Illustration:

Figure 1
We also update the model description with this Illustration

Related to https://github.com/neuropoly/totalspineseg/issues/50

### CODE REVIEW

**Comment by NathanMolinier** [comment]
I just moved from my old version to this PR to run the code and I got really bad results.
res_step2_output

The most probable reason is that my old models (step1 and step2) were used to run this PR. Is there a way to make sure that we are using the appropriate weights ? Maybe we could attach the release name to the weights: 
- `Dataset101_TotalSpineSeg_step1` --> `Dataset101_TotalSpineSeg_step1_rXXXXX` 
- `Dataset102_TotalSpineSeg_step2` --> `Dataset102_TotalSpineSeg_step2_rXXXXX`

**Comment by yw7** [comment]
> I just moved from my old version to this PR to run the code and I got really bad results. res_step2_output
> 
> The most probable reason is that my old models (step1 and step2) were used to run this PR. Is there a way to make sure that we are using the appropriate weights ? Maybe we could attach the release name to the weights:
> 
> * `Dataset101_TotalSpineSeg_step1` --> `Dataset101_TotalSpineSeg_step1_rXXXXX`
> * `Dataset102_TotalSpineSeg_step2` --> `Dataset102_TotalSpineSeg_step2_rXXXXX`

Yes!! Definitely we should,
We will do it in the next release.
Also I wanted to discuss what the best way to make the script update the model? Should we remove and update if out of date? Prompt the user? add another arguments to overwrite existing? Or simply add to the readme to run `rm -rf $TOTALSPINESEG_DATA/*`

**Comment by NathanMolinier** [comment]
> Also I wanted to discuss what the best way to make the script update the model? Should we remove and update if out of date? Prompt the user? add another arguments to overwrite existing? Or simply add to the readme to run `rm -rf $TOTALSPINESEG_DATA/*`

I personally think that we should not expect the user to know if he's using the right models. But these recent changes are also really particular because we are changing the outputs of the networks which should not happen too frequently. 

However, I still think that we should point to a specific release even inside the main code. This is already done when we download the weights so let's do what I suggested:
- Dataset101_TotalSpineSeg_step1 --> Dataset101_TotalSpineSeg_step1_rXXXXX
- Dataset102_TotalSpineSeg_step2 --> Dataset102_TotalSpineSeg_step2_rXXXXX

**Comment by yw7** [comment]
> > Also I wanted to discuss what the best way to make the script update the model? Should we remove and update if out of date? Prompt the user? add another arguments to overwrite existing? Or simply add to the readme to run `rm -rf $TOTALSPINESEG_DATA/*`
> 
> I personally think that we should not expect the user to know if he's using the right models. But these recent changes are also really particular because we are changing the outputs of the networks which should not happen too frequently.
> 
> However, I still think that we should point to a specific release even inside the main code. This is already done when we download the weights so let's do what I suggested:
> 
> * Dataset101_TotalSpineSeg_step1 --> Dataset101_TotalSpineSeg_step1_rXXXXX
> * Dataset102_TotalSpineSeg_step2 --> Dataset102_TotalSpineSeg_step2_rXXXXX
> 
> In the pyproject.toml:
> 
> ```
> [project.urls]
> homepage = "https://github.com/neuropoly/totalspineseg"
> repository = "https://github.com/neuropoly/totalspineseg"
> Dataset101_TotalSpineSeg_step1 = "https://github.com/neuropoly/totalspineseg/releases/download/r20240921/Dataset101_TotalSpineSeg_step1.zip"
> Dataset102_TotalSpineSeg_step2 = "https://github.com/neuropoly/totalspineseg/releases/download/r20240921/Dataset102_TotalSpineSeg_step2.zip"
> ```
> 
> Concerning the old weights I think that we can keep them.

Yess !! you definitely right! We should put the release name in the file mane.
The problem is that after extracting the model to results folder it is more complicates to know if it is in the right version, also if not, should we remove this release folder without prompting the user?
I think the easy way for us for now is to instruct the user to run `rm -rf $TOTALSPINESEG_DATA/nnUNet/{exports,results}` after update.
But we can discuss other strategies

**Comment by NathanMolinier** [comment]
Why can't we keep the release inside the name of the folder after extraction ? Also, we should not remove anything and keep all the previous weights

**Comment by NathanMolinier** [comment]
Update --> I removed the weights from the folder and got the same bad results

**Comment by yw7** [comment]
> rm -rf $TOTALSPINESEG_DATA/nnUNet/{exports,results}

Did you run `rm -rf $TOTALSPINESEG_DATA/nnUNet/{exports,results}` ?
you should remove both folders

**Comment by NathanMolinier** [comment]
I removed the full folder data

**Comment by NathanMolinier** [comment]
Solution: I needed to rerun `pip install -e .` to deal with the updated `pyproject.toml`
res_step2_output_tags

File: totalspineseg/utils/cpdir.py [hunk 1]
```diff
@@ -141,18 +150,42 @@ def cpdir_mp(
...
+    if not overwrite and dst_path.exists():
         return
 
-    # Make sure dst directory exists and copy
+    # Make sure dst directory exists
     dst_path.parent.mkdir(parents=True, exist_ok=True)
-    shutil.copy2(src_path, dst_path)
+
+    # Try copying up to num_tries times
+    for attempt in range(num_tries):
```
Comment (NathanMolinier) [inline totalspineseg/utils/cpdir.py:167]: If it doesn't work the first time, is possible that it will work after ?

Comment (yw7) [reply totalspineseg/utils/cpdir.py:167]: Yes
In network drive like in compute canada

File: totalspineseg/utils/extract_alternate.py [hunk 2]
```diff
@@ -15,6 +15,11 @@ def main():
     parser = argparse.ArgumentParser(
         description=' '.join(f'''
             Extract alternate labels from the segmentation.
+            This script extracts binary masks that include every other intervertebral disc (IVD).
```
Comment (NathanMolinier) [inline totalspineseg/utils/extract_alternate.py:18]:
```suggestion
            This script extracts binary masks that include every other intervertebral discs (IVD).
```

**Comment by NathanMolinier** [comment]
I noticed an issue with the labeling algorithm: here the vertebra C6 does not appear and C5 is used 2 times

Screenshot 2024-09-30 at 10 27 12

**Comment by yw7** [comment]
> I noticed an issue with the labeling algorithm: here the vertebra C6 does not appear and C5 is used 2 times
> 
> 
> 
> Screenshot 2024-09-30 at 10 27 12
> 
> 

Can you put here also the raw output of the 2 steps? Also if you can send me the raw output I'll investigate what happened

**Review by NathanMolinier — APPROVED** [review]
Seems to be working fine for the rest. The labeling issue has to be fixed.

**Pull request merged**